src/target/riscv/riscv-013.c

```diff
@@ -4810,73 +4908,96 @@ static int write_memory_progbuf(struct target *target, target_addr_t address,
 	/* Restore MSTATUS */
 	if (mstatus != mstatus_old)
 		if (register_write_direct(target, GDB_REGNO_MSTATUS, mstatus_old))
-			return ERROR_FAIL;
+			return MEM_ACCESS_FAILED;
 
 	if (execute_fence(target) != ERROR_OK)
-		return ERROR_FAIL;
+		return MEM_ACCESS_SKIPPED_FENCE_EXEC_FAILED;
 
-	return result;
+	return result == ERROR_OK ? MEM_ACCESS_OK : MEM_ACCESS_FAILED;
 }
 
-static int write_memory(struct target *target, target_addr_t address,
-		uint32_t size, uint32_t count, const uint8_t *buffer)
+static mem_access_result_t
+write_memory_sysbus(struct target *target, target_addr_t address,
+	uint32_t size, uint32_t count, const uint8_t *buffer)
 {
-	if (size != 1 && size != 2 && size != 4 && size != 8 && size != 16) {
-		LOG_TARGET_ERROR(target, "BUG: Unsupported size for memory write: %d", size);
-		return ERROR_FAIL;
-	}
+	riscv013_info_t *info = get_info(target);
+	mem_access_result_t skip_reason =
+		mem_should_skip_sysbus(target, address, size, 0, /* is_read = */ false);
+	if (skip_reason != MEM_ACCESS_OK)
+		return skip_reason;
 
+	/* TODO: write_memory_bus_* should return mem_access_result_t too*/
 	int ret = ERROR_FAIL;
-	RISCV_INFO(r);
-	RISCV013_INFO(info);
-
-	char *progbuf_result = "disabled";
-	char *sysbus_result = "disabled";
-	char *abstract_result = "disabled";
-
-	for (unsigned int i = 0; i < RISCV_NUM_MEM_ACCESS_METHODS; i++) {
-		int method = r->mem_access_methods[i];
-
-		if (method == RISCV_MEM_ACCESS_PROGBUF) {
-			if (mem_should_skip_progbuf(target, address, size, false, &progbuf_result))
-				continue;
-
-			ret = write_memory_progbuf(target, address, size, count, buffer);
+	uint64_t sbver = get_field(info->sbcs, DM_SBCS_SBVERSION);
+	if (sbver == 0)
+		ret = write_memory_bus_v0(target, address, size, count, buffer);
+	else if (sbver == 1)
+		ret = write_memory_bus_v1(target, address, size, count, buffer);
+	else
+		LOG_TARGET_ERROR(target,
+			"Unknown system bus version: %" PRIu64, sbver);
```

Comment:
1) If checkpatch allows the braces here, it'd help readability.

2) Also a blank line before and after the `if-else` cascade would help (especially to visually separate it from the below `if`).

```suggestion

	if (sbver == 0) {
		ret = write_memory_bus_v0(target, address, size, count, buffer);
	} else if (sbver == 1) {
		ret = write_memory_bus_v1(target, address, size, count, buffer);
	} else {
		LOG_TARGET_ERROR(target,
			"Unknown system bus version: %" PRIu64, sbver);
	}
			
```

Reply:
Addressed